api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -378,6 +387,324 @@ public T get() {
     }
   }
 
+  public static class WrappedStruct extends WrappedType {
+    private Iterable<WrappedType> fields;
+
+    public WrappedStruct(Type type, Iterable<WrappedType> fields) {
+      super(type);
+      this.fields = fields;
+    }
+
+    public Iterable<WrappedType> getFields() {
+      return fields;
+    }
+  }
+
+  public static class WrappedMap extends WrappedType {
+    private WrappedType key;
+    private WrappedType value;
+
+    public WrappedMap(Type type, WrappedType key, WrappedType value) {
+      super(type);
+      this.key = key;
+      this.value = value;
+    }
+
+    public WrappedType getKey() {
+      return key;
+    }
+
+    public WrappedType getValue() {
+      return value;
+    }
+  }
+
+  public static class WrappedList extends WrappedType {
+
+    private WrappedType list;
+
+    public WrappedList(Type type, WrappedType list) {
+      super(type);
+      this.list = list;
+    }
+
+    public WrappedType getList() {
+      return list;
+    }
+  }
+
+  /**
+   * POJO used to model an object describing a type alongside with a String representation of its name
+   * and a String representation of the nested structure of fields.
+   * The nested structure of fields is flattened and we use a dot delimited order for indicating precedence.
+   */
+  public static class WrappedType {
+
+    private String predecessor;
+    private String name;
+    private Type type;
+
+    public WrappedType(Type type) {
+      this.type = type;
+    }
+
+    public WrappedType setPredecessor(String predecessor) {
+      this.predecessor = predecessor;
+      return this;
+    }
+
+    public WrappedType setName(String name) {
+      this.name = name;
+      return this;
+    }
+
+    public String getPredecessor() {
+      return predecessor;
+    }
+
+    public String getName() {
+      return name;
+    }
+
+    public Type getType() {
+      return type;
+    }
+  }
+
+  /**
+   * Visitor for traversing two Iceberg schemas
+   *
+   * @param <T> the Java class returned by the visitor
+   */
+  public static class CompanionSchemaVisitor<T extends WrappedType> {
+    protected LinkedList<String> fields = Lists.newLinkedList();
+    protected LinkedList<Integer> fieldIds = Lists.newLinkedList();
+
+    public static <T extends WrappedType> List<T> schema(Schema schema, Schema companion, CompanionSchemaVisitor<T> visitor) {
+      return visitFields(schema.asStruct(), companion.asStruct(), visitor)
+          .stream()
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+    }
+
+    public static <T extends WrappedType> List<T> visit(Type left, Type right, CompanionSchemaVisitor<T> visitor,
```

Comment:
Why can visit only return a `WrappedType`? This seems too closely tied to its use.